examples/classic/distributed/TwoPhaseCommit/two_phase_commit.qnt

```diff
@@ -0,0 +1,137 @@
+
+  type Message =
+    | Abort
+    | Commit
+    | RMPrepared(str)
+    | RMCommitted(str)
+
+  var messages: Set[Message]
+
+  const resourceManagers: Set[str]
+
+  action init = all {
+    resourceManagerStates' = resourceManagers.mapBy(r => Working),
+    transactionManagerState' = Init,
+    preparedResourceManagers' = Set(),
+    messages' = Set(),
+  }
+
+  /* Transaction Manager actions */
+
+  action addPreparedResourceManager(resourceManager: str): bool = all {
+    transactionManagerState == Init,
+    messages.contains(RMPrepared(resourceManager)),
+    preparedResourceManagers' = preparedResourceManagers.union(Set(resourceManager))
+  }
+
+  action decideOnCommit = all {
+    transactionManagerState == Init,
+    preparedResourceManagers == resourceManagers,
+    transactionManagerState' = Done,
+    messages' = messages.union(Set(Commit))
+  }
+
+  action decideOnAbort = all {
+    transactionManagerState == Init,
+    transactionManagerState' = Done,
+    messages' = messages.union(Set(Abort))
+  }
+
+  /* Resource Manager actions */
+
+  action prepare(resourceManager: str): bool = all {
+    resourceManagerStates.get(resourceManager) == Working,
+    resourceManagerStates' = resourceManagerStates.set(resourceManager, Prepared),
+    messages' = messages.union(Set(RMPrepared(resourceManager)))
+  }
+
+  action abort(resourceManager: str): bool = all {
+    or {
+      // Resource manager chooses to abort
+      resourceManagerStates.get(resourceManager) == Working,
+      // Resource manager receives an abort message from the transaction manager
+      messages.contains(Abort),
+    },
+    resourceManagerStates' = resourceManagerStates.set(resourceManager, Aborted),
+    // There is no need to let others know that this resource manager aborted
+    messages' = messages
+  }
+
+  action commit(resourceManager: str): bool = all {
+    messages.contains(Commit),
+    resourceManagerStates' = resourceManagerStates.set(resourceManager, Committed),
+    messages' = messages.union(Set(RMCommitted(resourceManager)))
+  }
+
+  action step = any {
+    all {
+      nondet rm = resourceManagers.oneOf()
+      any { prepare(rm), abort(rm), commit(rm) },
+      transactionManagerState' = transactionManagerState,
+      preparedResourceManagers' = preparedResourceManagers,
+    },
+    all {
+      any { decideOnCommit, decideOnAbort },
+      preparedResourceManagers' = preparedResourceManagers,
+      resourceManagerStates' = resourceManagerStates,
+    },
+    all {
+      nondet rm = resourceManagers.oneOf()
+      addPreparedResourceManager(rm),
+      transactionManagerState' = transactionManagerState,
+      resourceManagerStates' = resourceManagerStates,
+      messages' = messages,
+    }
+  }
+
+  val consistencyInv = tuples(resourceManagers, resourceManagers).forall(((r1, r2)) =>
+    not(resourceManagerStates.get(r1) == Committed and resourceManagerStates.get(r2) == Aborted)
+  )
+}
+
+module two_phase_commit_3 {
```

Comment:
shall we extract this module into a separate file?

Reply:
I'd rather not. Most of our examples are in a single file, and adding another one increases the complexity in the examples dashboard. I think this is small enough to keep here for now.